// @fluentui/react-northstar/dist/commonjs/utils/positioner/Popper.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.Popper = void 0;

var _isEqual2 = _interopRequireDefault(require("lodash/isEqual"));

var _reactBindings = require("@fluentui/react-bindings");

var _reactComponentRef = require("@fluentui/react-component-ref");

var PopperJs = _interopRequireWildcard(require("@popperjs/core"));

var React = _interopRequireWildcard(require("react"));

var _isBrowser = require("../isBrowser");

var _getBoundary = require("./getBoundary");

var _getScrollParent = require("./getScrollParent");

var _positioningHelper = require("./positioningHelper");

/**
 * Memoize a result using deep equality. This hook has two advantages over
 * React.useMemo: it uses deep equality to compare memo keys, and it guarantees
 * that the memo function will only be called if the keys are unequal.
 * React.useMemo cannot be relied on to do this, since it is only a performance
 * optimization (see https://reactjs.org/docs/hooks-reference.html#usememo).
 *
 * Copied from https://github.com/apollographql/react-apollo/blob/master/packages/hooks/src/utils/useDeepMemo.ts.
 */
function useDeepMemo(memoFn, key) {
  var ref = React.useRef();

  if (!ref.current || !(0, _isEqual2.default)(key, ref.current.key)) {
    ref.current = {
      key: key,
      value: memoFn()
    };
  }

  return ref.current.value;
}
/** Checks if components was mounted the first time. */


function useFirstMount() {
  var isFirst = React.useRef(true);

  if (isFirst.current) {
    isFirst.current = false;
    return true;
  }

  return isFirst.current;
}
/** Executes useIsomorphicLayoutEffect during only updates. */


var useUpdateIsomorphicLayoutEffect = function useUpdateIsomorphicLayoutEffect(effect, deps) {
  var isFirstMount = useFirstMount();
  (0, _reactBindings.useIsomorphicLayoutEffect)(function () {
    if (!isFirstMount) {
      return effect();
    }
  }, deps);
};
/**
 * Popper relies on the 3rd party library [Popper.js](https://github.com/FezVrasta/popper.js) for positioning.
 */


var Popper = function Popper(props) {
  var align = props.align,
      children = props.children,
      enabled = props.enabled,
      flipBoundary = props.flipBoundary,
      userModifiers = props.modifiers,
      offset = props.offset,
      overflowBoundary = props.overflowBoundary,
      pointerTargetRef = props.pointerTargetRef,
      popperRef = props.popperRef,
      position = props.position,
      positionFixed = props.positionFixed,
      _props$positioningDep = props.positioningDependencies,
      positioningDependencies = _props$positioningDep === void 0 ? [] : _props$positioningDep,
      rtl = props.rtl,
      targetRef = props.targetRef,
      unstable_pinned = props.unstable_pinned;
  var proposedPlacement = (0, _positioningHelper.getPlacement)({
    align: align,
    position: position,
    rtl: rtl
  });
  var popperInstanceRef = React.useRef();
  var contentRef = React.useRef(null);
  var latestPlacement = React.useRef(proposedPlacement);

  var _React$useState = React.useState(proposedPlacement),
      computedPlacement = _React$useState[0],
      setComputedPlacement = _React$useState[1];

  var computedModifiers = useDeepMemo(function () {
    return [offset && {
      name: 'offset',
      options: {
        offset: rtl ? (0, _positioningHelper.applyRtlToOffset)(offset) : offset
      }
    }].concat(userModifiers);
  }, [offset, flipBoundary, overflowBoundary, userModifiers]);
  var createInstance = React.useCallback(function () {
    var reference = targetRef && (0, _reactComponentRef.isRefObject)(targetRef) ? targetRef.current : targetRef;

    if (!enabled || !reference || !contentRef.current) {
      return;
    }

    var handleUpdate = function handleUpdate(_ref) {
      var state = _ref.state;

      // PopperJS performs computations that might update the computed placement: auto positioning, flipping the
      // placement in case the popper box should be rendered at the edge of the viewport and does not fit
      if (state.placement !== latestPlacement.current) {
        latestPlacement.current = state.placement;
        setComputedPlacement(state.placement);
      }
    };

    var hasDocument = (0, _isBrowser.isBrowser)();
    var scrollParentElement = hasDocument ? (0, _getScrollParent.getScrollParent)(contentRef.current) : null;
    var hasScrollableElement = scrollParentElement ? scrollParentElement !== scrollParentElement.ownerDocument.body : false;
    var hasPointer = !!(pointerTargetRef && pointerTargetRef.current);
    var modifiers = [{
      name: 'flip',
      options: {
        flipVariations: true
      }
    },
    /**
     * unstable_pinned disables the flip modifier by setting flip.enabled to false; this
     * disables automatic repositioning of the popper box; it will always be placed according to
     * the values of `align` and `position` props, regardless of the size of the component, the
     * reference element or the viewport.
     */
    unstable_pinned && {
      name: 'flip',
      enabled: false
    },
    /**
     * When the popper box is placed in the context of a scrollable element, we need to set
     * preventOverflow.escapeWithReference to true and flip.boundariesElement to 'scrollParent'
     * (default is 'viewport') so that the popper box will stick with the targetRef when we
     * scroll targetRef out of the viewport.
     */
    hasScrollableElement && {
      name: 'flip',
      options: {
        boundary: 'clippingParents'
      }
    }, hasScrollableElement && {
      name: 'preventOverflow',
      options: {
        boundary: 'clippingParents'
      }
    }];
    var options = {
      placement: proposedPlacement,
      strategy: positionFixed ? 'fixed' : 'absolute',
      modifiers: [].concat(modifiers, computedModifiers, [
      /**
       * This modifier is necessary in order to render the pointer. Refs are resolved in effects, so it can't be
       * placed under computed modifiers. Deep merge is not required as this modifier has only these properties.
       */
      {
        name: 'arrow',
        enabled: hasPointer,
        options: {
          element: pointerTargetRef && pointerTargetRef.current
        }
      }, flipBoundary && {
        name: 'flip',
        options: {
          altBoundary: true,
          boundary: (0, _getBoundary.getBoundary)(contentRef.current, flipBoundary)
        }
      }, overflowBoundary && {
        name: 'preventOverflow',
        options: {
          altBoundary: true,
          boundary: (0, _getBoundary.getBoundary)(contentRef.current, overflowBoundary)
        }
      }, {
        name: 'onUpdate',
        enabled: true,
        phase: 'afterWrite',
        fn: handleUpdate
      }]).filter(Boolean),
      onFirstUpdate: function onFirstUpdate(state) {
        return handleUpdate({
          state: state
        });
      }
    };
    popperInstanceRef.current = PopperJs.createPopper(reference, contentRef.current, options);
  }, [contentRef, computedModifiers, enabled, flipBoundary, overflowBoundary, pointerTargetRef, positionFixed, proposedPlacement, targetRef, unstable_pinned]);
  var destroyInstance = React.useCallback(function () {
    if (popperInstanceRef.current) {
      popperInstanceRef.current.destroy();
      popperInstanceRef.current = null;
    }
  }, []);
  var scheduleUpdate = React.useCallback(function () {
    if (popperInstanceRef.current) {
      popperInstanceRef.current.update();
    }
  }, []);
  React.useImperativeHandle(popperRef, function () {
    return {
      updatePosition: scheduleUpdate
    };
  }, [scheduleUpdate]);
  (0, _reactBindings.useIsomorphicLayoutEffect)(function () {
    createInstance();
    return destroyInstance;
  }, [createInstance]);
  useUpdateIsomorphicLayoutEffect(scheduleUpdate, [].concat(positioningDependencies, [computedPlacement]));
  var child = typeof children === 'function' ? children({
    placement: computedPlacement,
    scheduleUpdate: scheduleUpdate
  }) : children;
  return child ? /*#__PURE__*/React.createElement(_reactComponentRef.Ref, {
    innerRef: contentRef
  }, React.Children.only(child)) : null;
};

exports.Popper = Popper;
Popper.defaultProps = {
  enabled: true,
  modifiers: [],
  positionFixed: false,
  positioningDependencies: []
};
